### PR TITLE
add -os=<os> switch

### PR DESCRIPTION
**Pull request by WalterBright**
Since my PRs to fix triples (they are currently unusable) are DOA,
https://github.com/dlang/dmd/pull/13486 https://github.com/dlang/dmd/pull/13484, I have given up on them.

In order to make cross-compilation work, only the operating system needs to be specified. This adds the support via the `-os=<os>` switch. Very easy.

### CODE REVIEW

**Review by RazvanN7 — APPROVED** [review]
Looks good to me. cc @ibuclaw @kinke

**Comment by RazvanN7** [comment]
@WalterBright is it ok to close #13486 and #13484 ?

**Review by thewilsonator — DISMISSED** [review]
Why does this need a separate switch? See also https://github.com/dlang/dmd/pull/13486#issuecomment-1003707861

Also why is this code not part of target.d.

**Comment by WalterBright** [comment]
> Why does this need a separate switch? See also #13486 (comment) Also why is this code not part of target.d.

It needs a separate switch because my changes to make this work with -target were rejected.

**Comment by WalterBright** [comment]
@RazvanN7 I figured on leaving them up for a while to see if someone else wants to carry on with fixing -target.

**Comment by WalterBright** [comment]
@thewilsonator it's not part of target.d so all the command line parsing goes in one place. Collecting all the command line parameters is different from checking them for consistency and validity. It would be fine if the validity checks were in target.d

**Comment by thewilsonator** [comment]
> It needs a separate switch because my changes to make this work with -target were rejected.

No, your changes to specify the components of the triple in an arbitrary order was rejected, and the other PR you appear to have abandoned for reasons of CI(?). 

Having `-target=os` was not part of the grounds for rejection and I think is reasonable.

>  Collecting all the command line parameters is different from checking them for consistency and validity. It would be fine if the validity checks were in target.d

What is the difference here? They all look to be validity checks to me.

**Comment by WalterBright** [comment]
Validity checks would be something like giving an error if `-fPIC` is used with `-os=windows`. All this code does is simply set the `params.os` field, it does not check whether it makes sense overall.

> your changes to specify the components of the triple in an arbitrary order was rejected

Not just that. I objected to the arguments about following a standard, because there apparently was no standard. It was also apparent that nobody was using it, which did not auger well for it being needed. And my goal is not to belabor all these points, but to get cross compiling working.

If you want to continue with `target=`, that's ok with me. But in the meantime, I'm going to get cross compiling working.

**Comment by WalterBright** [comment]
Once I get `-os=` working, it will make your task with fixing `-target=` much easier.

**Comment by thewilsonator** [comment]
> I objected to the arguments about following a standard, because there apparently was no standard.

The standard is _de facto_, i.e. what gcc, llvm (and by extension clang), the GNU tools have using for decades. If you want standard to mean there is a technical document that describes it, then you miss the forest for the trees in standard meaning what everyone (except for maybe Microsoft, don't know, don't care) does.

>  But in the meantime, I'm going to get cross compiling working.

and that doesn't work with `-target=os` why? This does not require a separate switch.

>  your task with fixing -target= much easier.

DMD is dead on OSX with ARM processors, going forward it will be dead completely on OSX. I don't develop with or on DMD at all because can't be built even with LDC.

**Comment by WalterBright** [comment]
> The standard is de facto, i.e. what gcc, llvm (and by extension clang), the GNU tools have using for decades. If you want standard to mean there is a technical document that describes it, then you miss the forest for the trees in standard meaning what everyone (except for maybe Microsoft, don't know, don't care) does.

The existing implementation of `-target` does not adhere to those conventions, it invents its own, as I pointed out in the PR. For example, the convention has a max of 4 arguments, `-target` has 5.

> and that doesn't work with -target=os why?

By rejecting out-of-order arguments, `-target=os` cannot work. The architecture comes first.

> I don't develop with or on DMD at all because can't be built even with LDC.

Then I am confused about what the test suite entitled `Ubuntu 18.04 x64, LDC` does.

I'd like to reiterate that I am not rejecting `-target`. I am not standing in the way of it. If you or someone else wants to get it to a working state according to your vision, please do so.

**Comment by thewilsonator** [comment]
> Then I am confused about what the test suite entitled Ubuntu 18.04 x64, LDC does.

Tests that DMD built with LDC passes, you missed the preceding sentence: DMD does not compile _at all_ with an aarch64 host.  Also FWIW the cirrus CI MacOS 10.x, 11.x, and 12.x jobs are all `x64` not `aarch64`.

> By rejecting out-of-order arguments, -target=os cannot work. The architecture comes first.

That is clearly and patently false, that has zero impact on the usability of that switch. If you can't remember the order, put it in `dmd.conf`.

> The existing implementation of -target does not adhere to those conventions, it invents its own, as I pointed out in the PR. For example, the convention has a max of 4 arguments, -target has 5.

It has 5 because otherwise it is impossible to target whatever clangs C++ library is on linux or windows, or glibc++ on OSX, etc. Sure we can make it adhere and not be a functional superset, and add yet another superfluous switch to specify the C++ environment to link against separately. It would be better than adding this superfluous switch.

If you're cross compiling and not supplying the entire triple, you're doing it wrong, plain and simple.  For example: windows host, compiling for linux. You  pass `-os=linux` and forget to supply the C environment, you'll target the MS C runtime (or even worse DMC runtime). Hopefully this fails at compile and not link time. Being able to independently supply the OS for cross compilation is bad UX because users _will_ forget to supply other critical information.

**Comment by ibuclaw** [comment]
> By rejecting out-of-order arguments, `-target=os` cannot work. The architecture comes first.
> 

That's not strictly true, as I think I said in the other PR, the OS has to be unambiguous though. i.e: `-target=mingw32` is unambiguously `i686`, `-target=mingw64` is unambiguously `x86_64`, `-target=msdos` is unambiguously `i386`, and so on.

**Comment by ibuclaw** [comment]
> By rejecting out-of-order arguments,

But, if you want to explicitly say what the architecture is, then the convention is to have it first. It's `-target=x86_64-darwin18`, and not `-target=darwin18-x86_64`. This is the crux of what's being rejected, and not `-target=os`.

**Comment by WalterBright** [comment]
@ibuclaw however, it *is* strictly true with the current implementation of target.

This whole triple thing, with no specification, seems to be a pile of contradictions and guesswork. I'm not opposed to it, and if someone else wants to fix it, that's fine. But nobody has stepped up in the last 7 months, and I need the cross compiler to work.

**Comment by WalterBright** [comment]
> That is clearly and patently false

It you try `-target=-windows`, you get "Error: unknown architecture `windows` for `-target`".

Ironically, my rejected PR fixed that problem.

> It has 5 because otherwise it is impossible to target whatever

Exactly. The current `-target` code is *inventing* a feature, not being *compatible* with a standard feature. The `+xxx` suffixes are another invention.

> That's not strictly true, as I think I said in the other PR, the OS has to be unambiguous though

It *is* strictly true with the current implementation of `-target`. My PR took advantage of the unambiguity of the arguments to allow any order.

> the convention is

Allowing them in any order doesn't violate any convention the user might want to adhere to.

**Comment by WalterBright** [comment]
> DMD does not compile at all with an aarch64 host.

I.e. an ARM hosted compiler. Of course it doesn't, because dmd is not a cross compiler. Something I am trying to make progress on with this PR.

If there are other reasons it does not work for aarch64, please let me know and I'll see about fixing them.

**Comment by thewilsonator** [comment]
> I.e. an ARM hosted compiler. Of course it doesn't, because dmd is not a cross compiler. Something I am trying to make progress on with this PR.

You misunderstand. The ARM problem is not that DMD does not function as a cross compiler, the problem is that DMD _fails compilation and cannot be built_ on an aarch64 host. I'm not concerned that DMD can't run, or produce code that runs, I can't even compile it.

>  If there are other reasons it does not work for aarch64, please let me know and I'll see about fixing them.

see https://github.com/dlang/dmd/pull/12764 https://github.com/dlang/dmd/pull/12818 for past attempts.
TL;DR, it mostly boils down to (hosts LDC's) `real` (and by extension DMD's deduced `real_t`) being `double` on apple silicon aarch64 (n.b. in the case of non-apple aarch64, real should be `fp128` but that's an entirely different problem). This requires software floating point for `fp80` real, which is only enabled for Windows (64?). 

We should really deprecate and remove support for `fp80` real, its old, not very useful and _very slow_, but again that is an issue for another time.

**Comment by WalterBright** [comment]
> We should really deprecate and remove support for fp80 real

Even better, write an emulator for 80 bit doubles. After all, even gcc and clang support 80 bit doubles.

**Comment by WalterBright** [comment]
> The ARM problem is not that DMD does not function as a cross compiler, the problem is that DMD fails compilation and cannot be built on an aarch64 host. I'm not concerned that DMD can't run, or produce code that runs, I can't even compile it.

If the problem is it doesn't have an 80 bit emulator, then we should fix that.

**Comment by thewilsonator** [comment]
> If the problem is it doesn't have an 80 bit emulator, then we should fix that.

Again _I don't care_ that DMD doesn't run on ARM (I believe that it will run under rosetta, haven't tried don't care), I care that it doesn't compile.
We _have_ soft fp80 (in `root/longdouble.d`), its only enabled for windows though and there is a baked in assumption that if `real != fp80` then we are targeting windows. 

This is a configuration problem, there is no technical restriction of capability here. 
Look at the linked PR in my previous comment.

**Comment by WalterBright** [comment]
`root/longdouble.d` is not an emulator. It executes x87 instructions to do the computations. It will not work on an Arm processor.

**Comment by thewilsonator** [comment]
Let me repeat myself again to be unequivocally clear: I do not care wether or not it can run, I care only that it complies.

**Comment by ibuclaw** [comment]
> > That is clearly and patently false
> 
> It you try `-target=-windows`, you get "Error: unknown architecture `windows` for `-target`".
> 

That looks wrong, it surely should be `-target=windows`.

Secondly, as a primer, the minimum that would be required is something like:

```
// Or: string.split('-').length == 1
if (!strstr(string, "-"))
{
    parseOS();
}
else
{
    parseArch();
    parseVendor();
    parseOS();
    parseRuntime();
} 
```

**Comment by WalterBright** [comment]
> it surely should be

I tried several combinations. All failed. Give it a try!

> the minimum that would be required is something like:

Never, ever, use `strstr()` for anything :-)

Anyhow, the existing implementation does not work for any notion of working, for any purpose. I was trying to fix it. My fixes were rejected. I'll leave making it work up to you (or anyone else who has a clear idea of what it should do).

**Comment by ibuclaw** [comment]
> > We should really deprecate and remove support for fp80 real
> 
> Even better, write an emulator for 80 bit doubles. After all, even gcc and clang support 80 bit doubles.

Native floating point is banned throughout the dmd front-end when building for gdc. Everything (float, double, real) goes through an emulated float type that matches size and endianess of the target.

Another thing that dmd does wrong is set floating point properties in dmd.target to ie `T.mant_dig` of the host.  The original values are long lost and now only get passed from binary to binary as a trojan virus.

**Comment by WalterBright** [comment]
> Soon the same will start being true of internal representations of integers too (replacing [du]integer_t with a struct of 64-bit integer pair to do 128-bit int at CTFE would be the starting point). DMD would be wise to do the same.

It's already being worked on.

> Another thing that dmd does wrong is set floating point properties in dmd.target to ie T.mant_dig of the host.

That works because it is not (yet) a cross compiler.

> The original values are long lost and now only get passed from binary to binary as a trojan virus.

Trojan virus? What?

BTW, all these complaints about dmd should exist in bugzilla. It's really off topic for this PR, as this PR is not part of the bug database.

**Comment by kinke** [comment]
> Native floating point is banned throughout the dmd front-end when building for gdc. Everything (float, double, real) goes through an emulated float type that matches size and endianess of the target.

I thought you/GDC was using some 160-bit type for `real_t`, regardless of the target, so has that changed lately?

LDC is using the host's `real`, except for Windows x86/x64, where it's using the `dmd.root.longdouble` x87 implementation. That's a well-known current cross-compilation issue, particularly when cross-compiling on x86 to non-Apple AArch64 with its quadruple precision `real`. Apple of course diverges and uses double-precision as C `long double`, so cross-compiling with LDC on x86 with x87 `real_t` isn't problematic, and shouldn't be for DMD either. I guess what does need work is the DMD backend, supporting double-precision `real` too, not x87 exclusively.

**Comment by ibuclaw** [comment]
> > The original values are long lost and now only get passed from binary to binary as a trojan virus.
> 
> Trojan virus? What?

Thompson Trojan.

**Comment by ibuclaw** [comment]
> > Native floating point is banned throughout the dmd front-end when building for gdc. Everything (float, double, real) goes through an emulated float type that matches size and endianess of the target.
> 
> I thought you/GDC was using some 160-bit type for `real_t`, regardless of the target, so has that changed lately?
> 

Around the time toPrec was added, I also started rounding intermediate values at certain CTFE checkpoints to whatever type is in flight, so while it may still spill over the precision of the target, it doesn't do so for long (however CTFE may still differ from runtime, so toPrec is still needed from time to time).

**Comment by RazvanN7** [comment]
@WalterBright @ibuclaw how do we proceed here?

**Comment by WalterBright** [comment]
@RazvanN7 doing cross compilation is dead in the water without this PR. So let's move forward with it.

**Comment by ibuclaw** [comment]
> @RazvanN7 doing cross compilation is dead in the water without this PR. So let's move forward with it.

Then kill -target and remove it from the 2.098 changelog.

**Comment by WalterBright** [comment]
I figured I'd leave `-target` be for a time if anyone wants to address it's problems. If nobody steps up, then I'll revert it.

**Comment by thewilsonator** [comment]
>doing cross compilation is dead in the water without this PR. 

_Only_ because you can't remember how triples work. `-target` sets all the relevant fields.

>  figured I'd leave -target be for a time if anyone wants to address it's problems. 

_There are no problems with `-target` that will not be fixed by getting cross compilation working by fixing other dmdisms_. I want to make that bluntly clear.

**Comment by WalterBright** [comment]
> Only because you can't remember how triples work.

Making such personal remarks is inappropriate here.

**Comment by thewilsonator** [comment]
Why do I bother?
